[wasm][debugger] Debug on firefox

thaystg opened the pull request:
It's working both on VS and on Firefox:
- step
- resume
- see locals
- callstack
- evaluate expressions
- show properties

To Do:
- open a PR on aspnetcore to enable debug on firefox https://github.com/dotnet/aspnetcore/blob/042217378e9737d2ab9d8332302bc9ccf46331b5/src/Components/Web.JS/src/Platform/Mono/MonoDebugger.ts#L3

![MicrosoftTeams-image](https://user-images.githubusercontent.com/4503299/142430850-d15d0d3f-8d3c-42a4-b9e3-3a32015cfbbd.png)

![MicrosoftTeams-image2](https://user-images.githubusercontent.com/4503299/142430865-380277d7-c1c1-4daf-a56a-6eb16bda6dee.png)

thaystg committed:
First compiling and working version just proxying messages.

thaystg committed:
almost working, already showing the cs files

thaystg committed:
Merge branch 'dotnet:main' into debug_on_firefox

thaystg committed:
Working on firefox.

thaystg committed:
merge

thaystg committed:
Use internal e not public.

thaystg committed:
Merge branch 'dotnet:main' into debug_on_firefox

thaystg committed:
merge

thaystg committed:
Debugging on firefox working.

thaystg committed:
Merge remote-tracking branch 'origin/main' into debug_on_firefox

thaystg committed:
Working after the merge

thaystg committed:
Keep the TcpListener open and use a random port.

thaystg committed:
Show null value.

thaystg committed:
- Show JS Callstack
- Skip properties
- Get array value from evaluateAsyncJS and not use the preview value.

thaystg committed:
Merge branch 'main' into debug_on_firefox

thaystg committed:
Fix compilation

thaystg committed:
Infrastructure to run debugger tests.

thaystg committed:
Merge branch 'main' into debug_on_firefox

thaystg commented:
@radical when you have time, can you please start the review?

thaystg committed:
fix merge

thaystg committed:
run test.

thaystg committed:
Skipping tests that are not passing on Firefox.

thaystg committed:
Skipping tests that are not passing on Firefox.

thaystg committed:
Passing 13 steppingtests.

thaystg committed:
Passing 13 steppingtests.

thaystg committed:
Passing 13 steppingtests.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/FirefoxMonoProxy.cs:
```diff
@@ -0,0 +1,910 @@
+                    }
+                case "source":
+                    {
+                        if (args["to"].Value<string>().StartsWith("dotnet://"))
```
This check isn't needed, as `OnGetScriptSource` will check as the first thing.

So, this can be similar to MonoProxy:
```
                        string script = args?["scriptId"]?.Value<string>();
                        return await OnGetScriptSource(id, script, token);
```

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/FirefoxMonoProxy.cs:
```diff
@@ -0,0 +1,910 @@
+                    }
+                case "getBreakableLines":
+                    {
+                        if (args["to"].Value<string>().StartsWith("dotnet://"))
```
Same for this. `OnGetBreakableLines` checks the SourceId as the first thing, so this check isn't needed.

thaystg committed:
Failed:     0, Passed:    39, Skipped:   203, Total:   242, Duration: 5 m 6 s - DebuggerTestSuite.dll (net6.0)

thaystg committed:
Failed:     0, Passed:    66, Skipped:   195, Total:   261, Duration: 9 m 29 s - DebuggerTestSuite.dll (net6.0)

thaystg committed:
Merge remote-tracking branch 'origin/main' into debug_on_firefox

thaystg committed:
Using ConditionalTheory and ConditionalFact implemented by @radical.

thaystg committed:
Fixing side effect.

thaystg committed:
Implemented conditional breakpoints.
Failed:     0, Passed:    74, Skipped:   189, Total:   263, Duration: 8 m 41 s - DebuggerTestSuite.dll (net6.0)

thaystg committed:
Fix special characters and pointers.

Failed:     0, Passed:   116, Skipped:   177, Total:   293

thaystg committed:
Merge remote-tracking branch 'origin/main' into HEAD

thaystg committed:
Fix merge

thaystg committed:
Run debugger-tests on firefox using codespace

thaystg committed:
Starting firefox correctly not stopping in the breakpoint yet.

thaystg committed:
Remove unnecessary change

thaystg committed:
Fix pause behavior (now showing correctly, pause on breakpoint, pause while stepping)
Start implementing evaluate expressions, working correctly on VSCode.

thaystg committed:
Merge branch 'debug_on_firefox' of github.com:thaystg/runtime into debug_on_firefox

thaystg committed:
Fix local tests.

thaystg committed:
Fix missing )

thaystg committed:
Passing 190 tests, evaluate expressions working.

thaystg committed:
Merge remote-tracking branch 'origin/main' into debug_on_firefox

thaystg committed:
Remove Task.Delays.
Move some attributes from FirefoxMonoProxy to FirefoxExecutionContext.

thaystg committed:
Fix container creation

thaystg committed:
Trying to run firefox tests on CI.

thaystg committed:
Moving file to the right place.

thaystg committed:
Trying to run debugger-tests using firefox on CI.

thaystg commented:
/azp run runtime-extra-platforms

thaystg committed:
fixing path

thaystg committed:
Missing url to download firefox on helix.

thaystg commented:
/azp run runtime-wasm

thaystg committed:
On run the tests only on linux.

thaystg committed:
Trying to download firefox on helix.

thaystg committed:
fix error on helix-wasm.targets.

thaystg committed:
trying to fix ci

thaystg committed:
trying to install firefox on helix.

thaystg committed:
Fixing firefox path

thaystg committed:
Fix debugger tests on firefox

thaystg committed:
fixing profile path

thaystg committed:
Install libdbus-glib-1-2 on docker and on codespace

thaystg committed:
Merge remote-tracking branch 'thaystg/main' into pr/thaystg/61776

thaystg committed:
Trying to run using firefox on CI

lewing commented on src/mono/wasm/debugger/BrowserDebugProxy/DevToolsHelper.cs:
```diff
@@ -149,6 +151,89 @@ public static Result FromJson(JObject obj)
+                    if (arrayItem.Name != "length")
+                        ret.Add(arrayItem.Value["value"]);
+                }
+                o = JObject.FromObject(new
+                {
+                    result = JObject.FromObject(new
+                    {
+                        value = ret
+                    })
+                });
```
```suggestion
                o = JObject.FromObject(new
                {
                    result = new
                    {
                        value = ret
                    }
                });
```
You can nest anonymous objects rather than calling FromObject multiple times

lewing reviewed (commented):
Looking good, one comment while I keep reading

thaystg committed:
update docker image

thaystg committed:
Merge remote-tracking branch 'origin/main' into debug_on_firefox

thaystg committed:
Adding more messages to see errors on CI

thaystg committed:
Trying to make it work on CI

thaystg committed:
Real test on CI

thaystg committed:
Trying to use the firefox machine only to run firefox tests
Retrying connection to Proxy
Remove extra messages added to help to fix CI

thaystg committed:
Fix CI

thaystg committed:
Fix CI

thaystg committed:
Fix CI.

thaystg committed:
Remove unnecessary changes.

thaystg committed:
Using machine with sudo installed

thaystg committed:
Addressing @lewing comments

thaystg committed:
Merge remote-tracking branch 'origin/main' into debug_on_firefox

thaystg committed:
Fix run tests on codespace
Using image with python3

thaystg committed:
Use default image to build and new image only to run firefox tests

thaystg committed:
Fix unrelated change

thaystg committed:
Fix ci

thaystg committed:
check python version

thaystg committed:
Print python versions

thaystg committed:
Using image with PIP installed

thaystg committed:
Using image with pip updated

thaystg committed:
Remove unrelated changes
Increase time to wait for firefox to be ready

thaystg committed:
Trying to fix evaluate tests.

thaystg committed:
Fix evaluateoncallframe tests

thaystg committed:
Trying to fix evaluation tests.

thaystg committed:
trying to fix evaluateoncallframetests

thaystg committed:
fiz evaluateoncallframetests

thaystg committed:
Trying to kill firefox to avoid errors.

thaystg committed:
Trying to fix EvaluateOnCallFrameTests

thaystg committed:
Fix CI

thaystg committed:
Remove failing test

thaystg committed:
Fix misctests

thaystg committed:
Fix other build errors.

thaystg committed:
Trying to fix CI.

thaystg committed:
Fix CI

thaystg committed:
Remove unecessary message.

thaystg commented:
/azp run runtime-wasm

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DevToolsProxy.cs:
```diff
@@ -388,4 +398,48 @@ protected void Log(string priority, string msg)
             }
         }
     }
+    public class FirefoxProxyServer
```
- put new classes in separate files.
- use file-scoped namespaces in them, `namespaces foo;`
- Use `#nullable enable`

thaystg replied:
done.

radical commented on src/tests/BuildWasmApps/Wasm.Debugger.Tests/Wasm.Debugger.Tests.csproj:
```diff
@@ -37,10 +37,22 @@
-                         Include="export CHROME_PATH_FOR_DEBUGGER_TESTS=$HELIX_CORRELATION_PAYLOAD/chrome-linux/chrome" />
-      <RunScriptCommands Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(OS)' == 'Windows_NT'"
-                         Include="set CHROME_PATH_FOR_DEBUGGER_TESTS=%HELIX_CORRELATION_PAYLOAD%\chrome-win\chrome.exe" />
+      <RunScriptCommands Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(OS)' != 'Windows_NT' and '$(RunInFirefox)'!='1'"
+                         Include="export BROWSER_PATH_FOR_DEBUGGER_TESTS=$HELIX_CORRELATION_PAYLOAD/chrome-linux/chrome" />
+      <RunScriptCommands Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(OS)' == 'Windows_NT' and '$(RunInFirefox)'!='1'"
+                         Include="set BROWSER_PATH_FOR_DEBUGGER_TESTS=%HELIX_CORRELATION_PAYLOAD%\chrome-win\chrome.exe" />
+      <RunScriptCommands Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(OS)' != 'Windows_NT' and '$(RunInFirefox)'=='1'"
+                         Include="export BROWSER_PATH_FOR_DEBUGGER_TESTS=$HELIX_CORRELATION_PAYLOAD/firefox/firefox/firefox/firefox" />
+      <RunScriptCommands Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(OS)' != 'Windows_NT' and '$(RunInFirefox)'=='1'"
+                         Include="rm -rf /tmp/profile" />                         
+      <RunScriptCommands Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(OS)' != 'Windows_NT' and '$(RunInFirefox)'=='1'"
+                         Include="mkdir /tmp/profile" />
+      <RunScriptCommands Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(OS)' != 'Windows_NT' and '$(RunInFirefox)'=='1'"
+                         Include="echo 'user_pref(&quot;devtools.chrome.enabled&quot;, true)%3B' >> /tmp/profile/prefs.js" />
+      <RunScriptCommands Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(OS)' != 'Windows_NT' and '$(RunInFirefox)'=='1'"
+                         Include="echo 'user_pref(&quot;devtools.debugger.remote-enabled&quot;, true)%3B' >> /tmp/profile/prefs.js" />                         
+      <RunScriptCommands Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(OS)' != 'Windows_NT' and '$(RunInFirefox)'=='1'"
+                         Include="echo 'user_pref(&quot;devtools.debugger.prompt-connection&quot;, false)%3B' >> /tmp/profile/prefs.js" />
```
```suggestion
    <ItemGroup Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(RunInFirefox)' != '1'">
      <RunScriptCommands Condition="'$(OS)' != 'Windows_NT'" Include="export BROWSER_PATH_FOR_DEBUGGER_TESTS=$HELIX_CORRELATION_PAYLOAD/chrome-linux/chrome" />
      <RunScriptCommands Condition="'$(OS)' == 'Windows_NT'" Include="set BROWSER_PATH_FOR_DEBUGGER_TESTS=%HELIX_CORRELATION_PAYLOAD%\chrome-win\chrome.exe" />
    </ItemGroup>

    <ItemGroup Condition="'$(ContinuousIntegrationBuild)' == 'true' and '$(RunInFirefox)' == '1'">
      <RunScriptCommands Condition="'$(OS)' != 'Windows_NT'" Include="export BROWSER_PATH_FOR_DEBUGGER_TESTS=$HELIX_CORRELATION_PAYLOAD/firefox/firefox/firefox/firefox" />
      <RunScriptCommands Condition="'$(OS)' != 'Windows_NT'" Include="rm -rf /tmp/profile" />
      <RunScriptCommands Condition="'$(OS)' != 'Windows_NT'" Include="mkdir /tmp/profile" />
      <RunScriptCommands Condition="'$(OS)' != 'Windows_NT'" Include="echo 'user_pref(&quot;devtools.chrome.enabled&quot;, true)%3B' >> /tmp/profile/prefs.js" />
      <RunScriptCommands Condition="'$(OS)' != 'Windows_NT'" Include="echo 'user_pref(&quot;devtools.debugger.remote-enabled&quot;, true)%3B' >> /tmp/profile/prefs.js" />
      <RunScriptCommands Condition="'$(OS)' != 'Windows_NT'" Include="echo 'user_pref(&quot;devtools.debugger.prompt-connection&quot;, false)%3B' >> /tmp/profile/prefs.js" />
    </ItemGroup>

```

thaystg committed:
Update src/tests/BuildWasmApps/Wasm.Debugger.Tests/Wasm.Debugger.Tests.csproj

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Addressing @radical comments

thaystg committed:
Merge branch 'main' into debug_on_firefox

thaystg committed:
Merge error while accept @radical suggestion

thaystg committed:
Merge error while accept @radical suggestion

radical commented on eng/pipelines/common/templates/wasm-debugger-tests-firefox.yml:
```diff
@@ -0,0 +1,42 @@
+parameters:
```
1. Instead of a new template, add a `browser` parameter to `wasm-debugger-tests.yml`, with a default value of `chrome`
2. In that, pass `/p:RunInFirefox=1` if the `browser == firefox`
3. And use that for adding the new Firefox jobs

radical replied:
Something like https://github.com/radical/runtime/commit/655c0f17241c571fd68f0983a1c492be7a575ecf

thaystg replied:
done.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs:
```diff
@@ -138,6 +138,24 @@ public bool TryResolve(DebugStore store)
         }
 
+        public bool CompareRequest(JObject req)
+        {
+            if (this.request["url"].Value<string>() == req["url"].Value<string>() &&
+                this.request["lineNumber"].Value<int>() == req["lineNumber"].Value<int>() &&
+                this.request["columnNumber"].Value<int>() == req["columnNumber"].Value<int>())
+                return true;
+            return false;
+        }
```
```suggestion
          => this.request["url"].Value<string>() == req["url"].Value<string>() &&
                this.request["lineNumber"].Value<int>() == req["lineNumber"].Value<int>() &&
                this.request["columnNumber"].Value<int>() == req["columnNumber"].Value<int>();
```

thaystg committed:
Update src/mono/wasm/debugger/BrowserDebugProxy/DebugStore.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestSuite.csproj:
```diff
@@ -5,17 +5,21 @@
+    <DefineConstants Condition="'$(RunInFirefox)'!='1'">$(DefineConstants);RUN_IN_CHROME</DefineConstants>
     <VersionsPropsFile>$(MSBuildThisFileDirectory)..\..\BrowsersForTesting.props</VersionsPropsFile>
     <BrowserHost Condition="$([MSBuild]::IsOSPlatform('windows'))">windows</BrowserHost>
     <InstallChromeForDebuggerTests Condition="'$(InstallChromeForDebuggerTests)' == '' and '$(ContinuousIntegrationBuild)' != 'true' and Exists('/.dockerenv')">true</InstallChromeForDebuggerTests>
```
```suggestion
    <InstallChromeForDebuggerTests Condition="'$(InstallChromeForDebuggerTests)' == '' and '$(RunInFirefox)' != '1' and '$(ContinuousIntegrationBuild)' != 'true' and Exists('/.dockerenv')">true</InstallChromeForDebuggerTests>
```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestSuite.csproj:
```diff
@@ -5,17 +5,21 @@
-    <ChromeStampFile>$(ChromeStampDir).install-chrome-$(ChromiumRevision).stamp</ChromeStampFile>
+    <BrowserStampDir>$(ArtifactsBinDir)DebuggerTestSuite\</BrowserStampDir>
+    <ChromeStampFile>$(BrowserStampDir).install-chrome-$(ChromiumRevision).stamp</ChromeStampFile>
+    <FirefoxDir>$(ArtifactsBinDir)DebuggerTestSuite\firefox\</FirefoxDir>    
```
```suggestion
    <FirefoxDir>$(ArtifactsBinDir)DebuggerTestSuite\firefox\</FirefoxDir>
```

thaystg committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

thaystg committed:
Addressing @radical comments

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/DevToolsQueue.cs:
```diff
@@ -53,4 +69,33 @@ public bool TryPumpIfCurrentCompleted(CancellationToken token, [NotNullWhen(true
         }
     }
+
+    internal class DevToolsQueueFirefox : DevToolsQueueBase
```
split the top level classes into separate files.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/FirefoxMonoProxy.cs:
```diff
@@ -0,0 +1,1118 @@
+
+namespace Microsoft.WebAssembly.Diagnostics
+{
+    internal class FirefoxExecutionContext : ExecutionContext
```
same here, split the top level classes into separate files

radical committed:
Abort the tcp connection if the proxy throws an exception

radical committed:
Refactor a bit

radical committed:
Use more compile time checks for chrome vs firefox

radical commented:
/azp run runtime-wasm

radical committed:
fix pipeline

radical committed:
Make debugger job names unique by including the browser

radical commented:
/azp run runtime-wasm

radical committed:
Merge remote-tracking branch 'origin/main' into debug_on_firefox

radical committed:
fix runtime-wasm pipeline

radical commented:
/azp run runtime-wasm

radical committed:
fix firefox ci job

radical commented:
/azp run runtime-wasm

radical committed:
split into more files

radical committed:
Merge remote-tracking branch 'origin/main' into debug_on_firefox

radical committed:
cleanup

radical committed:
Add support for running chrome, and firefox tests in the same job

radical committed:
fix yml

radical committed:
fix build

radical commented:
/azp run runtime-wasm

radical committed:
fix build

radical commented:
/azp run runtime-wasm

radical committed:
fix windows build

radical commented:
/azp run runtime-wasm

thaystg committed:
Don't delete profile folder nor pkill firefox

thaystg committed:
Delete and create a new profile folder for each execution

radical committed:
fix helix command line

radical committed:
[wasm][debugger] Fix tests broken on 'main'

This test broke because it was checking for the number of members on
`System.TimeSpan`, and that changed with
https://github.com/dotnet/runtime/pull/67666 , which added new members
like `TotalNanoseconds`.

The test shouldn't depend on this number anyway, so remove that.

```
  Failed DebuggerTests.MiscTests.InspectLocalsForToStringDescriptions(line: 137, col: 12, method_name: "MethodWithLocalsForToStringTest", call_other: False, invoke_async: False) [758 ms]
  Error Message:
   [ts_props] Number of fields don't match, Expected: 12, Actual: 16
Expected: True
Actual:   False
  Stack Trace:
     at DebuggerTests.DebuggerTestBase.CheckProps(JToken actual, Object exp_o, String label, Int32 num_fields) in /_/src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestBase.cs:line 800
   at DebuggerTests.DebuggerTestBase.CompareObjectPropertiesFor(JToken locals, String name, Object o, String label, Int32 num_fields) in /_/src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestBase.cs:line 908
   at DebuggerTests.MiscTests.InspectLocalsForToStringDescriptions(Int32 line, Int32 col, String method_name, Boolean call_other, Boolean invoke_async) in /_/src/mono/wasm/debugger/DebuggerTestSuite/MiscTests.cs:line 559
```

radical committed:
Merge remote-tracking branch 'origin/main' into debug_on_firefox

radical committed:
wip

radical committed:
big refactor

radical committed:
chrome runs

radical committed:
ff runs

radical committed:
ff runs

radical committed:
Merge remote-tracking branch 'origin/main' into debug_on_firefox

radical committed:
cleanup

radical committed:
cleanup

radical committed:
cleanup

radical committed:
change console verbosity to info, for proxy

radical committed:
More refactoring

radical committed:
Merge remote-tracking branch 'origin/main' into thays/debug_on_firefox

radical committed:
Merge remote-tracking branch 'origin/main' into thays/debug_on_firefox

radical committed:
More refactoring, and fix some issues with connections, and other
cleanup

radical committed:
some cleanup

radical committed:
fix file name

radical committed:
Improve cleanup after tests

radical committed:
Merge remote-tracking branch 'origin/main' into upstream-debug_on_firefox

radical committed:
Merge remote-tracking branch 'origin/main' into upstream-debug_on_firefox

radical committed:
some refactoring, fixing some hangs, faster failures etc

radical committed:
Merge remote-tracking branch 'origin/main' into thays/debug_on_firefox

radical commented:
/azp run runtime-staging

radical committed:
Fix BrowserCrash test for chrome

radical committed:
Merge remote-tracking branch 'origin/main' into thays/debug_on_firefox

radical committed:
fix up logging

radical committed:
Improve error handling for the proxy running independently

radical committed:
fix debugging from vscode

radical committed:
proxy host: add --log-path for logs

radical committed:
support canceling for the proxy host too, and distinguish different instances of the proxy

thaystg committed:
Merge remote-tracking branch 'origin/main' into debug_on_firefox

thaystg committed:
Fix debugger after refreshing the debugged page.

thaystg committed:
Fixing chrome debugging.

thaystg committed:
Fix startup to work on chrome and also on firefox.

ilonatommy commented:
/azp run runtime-staging
